examples/godocbot/pkg/controller/pullrequest/github_syncer.go

```diff
@@ -0,0 +1,223 @@
+	ghClient *github.Client
+}
+
+func (r *pullRequestCommitIDReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+
+	log.Printf("reconciling request '%v'", request.NamespacedName)
+	// Fetch PullRequest object
+	pr := &v1alpha1.PullRequest{}
+	err := r.Client.Get(ctx, request.NamespacedName, pr)
+	if errors.IsNotFound(err) {
+		log.Printf("Could not find PullRequest %v.\n", request)
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		log.Printf("Could not fetch PullRequest %v for %+v\n", err, request)
+		return reconcile.Result{}, err
+	}
+
+	if pr.Spec.CommitID != "" {
+		// We already know the commit-id for the PR, so no need to update the commitID
+		return reconcile.Result{}, nil
+	}
+
+	// Looks like this is a fresh PR, so lets determine the latest commitID
+	prinfo, err := parsePullRequestURL(pr.Spec.URL)
+	if err != nil {
+		log.Printf("error in parsing the request, ignoring this pr %v err %v", request.NamespacedName, err)
+		return reconcile.Result{}, nil
+	}
+
+	log.Printf("fetching commit id for the PR: %v", prinfo)
+	ghPR, _, err := r.ghClient.PullRequests.Get(context.Background(), prinfo.org, prinfo.repo, int(prinfo.pr))
+	if err != nil {
+		log.Printf("error fetching PR details from github: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// deep copy ? check if it is still required with pkg/cache or client ?
+	prCopy := pr.DeepCopy()
+	prCopy.Spec.CommitID = ghPR.Head.GetSHA()
+	err = r.Client.Update(context.Background(), prCopy)
+	if err != nil {
+		log.Printf("error updating PR github: %v", err)
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+// Start periodically syncs PRs in k8s with their commitID in Github.
+func (gs *GithubSyncer) Start(stop <-chan struct{}) {
+	ticker := time.NewTicker(gs.syncInterval)
+	for {
+		select {
+		case <-stop:
+			/* we got signalled */
+			return
+		case <-ticker.C:
+			gs.syncPullRequests()
+		}
+	}
+}
+
+// syncPullRequests performs the following:
+//  - fetches all the PRs registered in K8s
+//  - Organize them by orgs/repo so that batch calls can be made
+//  - Fetches details of PRs from Github and updates the commitID of the PRs in
+//  k8s if required.
+// TODO(droot): delete the PRs from K8s if closed in Github.
+func (gs *GithubSyncer) syncPullRequests() {
+	prList := &v1alpha1.PullRequestList{}
+	// get pull requests in all namespaces
+	err := gs.mgr.GetClient().List(context.Background(), &client.ListOptions{Namespace: ""}, prList)
+	if err != nil {
+		log.Printf("error fetching all the PRs from k8s: %v", err)
+		return
+	}
+
+	// TODO(droot): simplify the following loop
+	orgs := pullRequestByRepoAndOrg(prList)
+	for org, repos := range orgs {
+		for repo, prs := range repos {
+			ghPRs, _, err := gs.ghClient.PullRequests.List(context.Background(), org, repo, nil)
+			if err != nil {
+				log.Printf("cannont get PR list from GH: %v", err)
+				continue
+			}
+			for _, ghPR := range ghPRs {
```

Comment:
Consider creation a separate pre-processing function that creates a map[int32]PR